Toolbox/ArcSDM.pyt

```diff
@@ -784,6 +786,163 @@ class CalculateWeights(object):
         return
 
 
+class CosineSimilarityIndex(object):
+    def __init__(self):
+        self.label = "Cosine Similarity Index (CSI)"
+        self.description = (
+            "Compute Cosine Similarity Index for labelled data and evidence data. "
+            "Outputs: (1) pairwise CSI matrix for labelled samples (CSV), "
+            "(2) CSI matrix between class centroids (CSV), "
+            "(3) CSI for evidence vectors (CSV) or a raster per class label."
+        )
+        self.category = f"{TS_EXPLORATORY_DATA_ANALYSIS}"
+        self.canRunInBackground = True
+
+    def getParameterInfo(self):
+
+        # --- Labelled data ---
+        param_labelled_data_path = arcpy.Parameter(
+            displayName="Labelled data path (CSV/TXT or Feature Class)",
+            name="labelled_path",
+            datatype= ["DEFile", "DEFeatureClass"],
+            parameterType="Required",
+            direction="Input",
+        )
+        
+        param_csv_no_data_sentinel = arcpy.Parameter(
+            displayName="CSV/TXT NoData sentinel (numeric, optional)",
+            name="csv_nodata",
+            datatype="GPDouble",
+            parameterType="Optional",
+            direction="Input",
+        )
+        
+        param_label_field_names = arcpy.Parameter(
+            displayName="Label field names",
+            name="label_field_names",
+            datatype="Field",
+            parameterType="Required",
+            direction="Input",
+            multiValue=True,
+        )
+        param_label_field_names.parameterDependencies = [param_labelled_data_path.name]
+
+        param_feature_field_names = arcpy.Parameter(
+            displayName="Field names (include coordinates)",
+            name="feature_field_names",
+            datatype="Field",
+            parameterType="Optional",
+            direction="Input",
+            multiValue=True,
+        )
+        param_feature_field_names.parameterDependencies = [param_labelled_data_path.name]
+
+        param_evidence_source_type = arcpy.Parameter(
+            displayName="Evidence source type",
+            name="evidence_type",
+            datatype="GPString",
+            parameterType="Required",
+            direction="Input",
+        )
+        param_evidence_source_type.filter.list = ["Raster"]
+        param_evidence_source_type.value = "Raster"
+
+        param_evidence_rasters = arcpy.Parameter(
+            displayName="Evidence rasters (same size/extent/projection)",

```

Comment:
Even though the displayName does tell the user what is expected, it doesn't warn about the potentially serious consequences. If the raster dimensions don't match, it can result in deletion of data. The output directory selected for saving the CSI rasters seems to get deleted if the samples.csv and centroids.csv paths were set to be within the output dir. If that directory contained something else as well, it would be gone. 

Something is not getting handled in the CSI raster creation step most likely. Not sure what the fix is, but we should try to fix the root cause.